src/main.rs

```diff
@@ -202,24 +214,14 @@ fn flashing_progress() -> flashing::FlashProgress {
     })
 }
 
-/// Read stack-pointer and reset-handler-address from the vector table.
-///
-/// Assumes that the target was reset-halted.
-///
-/// Returns `(stack_start: u32, reset_fn_address: u32)`
-fn analyze_vector_table(core: &mut Core) -> anyhow::Result<(u32, u32)> {
-    let stack_start = core.read_core_reg::<u32>(SP)?;
-    let reset_address = cortexm::set_thumb_bit(core.read_core_reg::<u32>(PC)?);
-    Ok((stack_start, reset_address))
-}
-
-fn start_program(core: &mut Core, elf: &Elf) -> anyhow::Result<()> {
-    log::debug!("starting device");
+fn attach_to_program(core: &mut Core, elf: &Elf) -> anyhow::Result<()> {
+    log::debug!("attaching to program");
 
     match (core.available_breakpoint_units()?, elf.rtt_buffer_address()) {
         (0, Some(_)) => bail!("RTT not supported on device without HW breakpoints"),
         (0, None) => log::warn!("device doesn't support HW breakpoints; HardFault will NOT make `probe-run` exit with an error code"),
-        (_, Some(rtt_buffer_address)) => set_rtt_to_blocking(core, elf.main_fn_address(), rtt_buffer_address)?,
+        (_, Some(rtt_buffer_address)) =>
+            set_rtt_blocking_mode(core, rtt_buffer_address, true)?,
```

Comment:
Don't we need to halt the program at the beginning of main? Before this PR this was done in `set_rtt_to_blocking`.

Reply:
Is there a reason it needs to be? I couldn't find one and that's why I removed the check that was in `set_rtt_to_blocking`.

Reply:
I assume (assume!) that we want to wait for potential startup code to finish before we mess around with the micro-controller. Runtime crates like `cortex_m_rt` ([link](https://docs.rs/cortex-m-rt/latest/cortex_m_rt/)) run some code before `main` and also allow users to specify custom code to run before `main`.